util/winutil/authenticode/authenticode_windows.go

```diff
@@ -0,0 +1,425 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package authenticode
+
+import (
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"path/filepath"
+	"strings"
+	"unsafe"
+
+	"github.com/dblohm7/wingoes"
+	"golang.org/x/sys/windows"
+)
+
+var (
+	// ErrSigNotFound is returned if no authenticode signature could be found.
+	ErrSigNotFound = errors.New("authenticode signature not found")
+	// ErrUnexpectedCertSubject is wrapped with the actual cert subject and
+	// returned when the binary is signed by a different subject than expected.
+	ErrUnexpectedCertSubject = errors.New("unexpected cert subject")
+	errCertSubjectNotFound = errors.New("cert subject not found")
+	errCertSubjectDecodeLenMismatch = errors.New("length mismatch while decoding cert subject")
+)
+
+const (
+	_CERT_STRONG_SIGN_OID_INFO_CHOICE = 2
+	_CMSG_SIGNER_CERT_INFO_PARAM      = 7
+	_MSI_INVALID_HASH_IS_FATAL        = 1
+	_TRUST_E_NOSIGNATURE              = wingoes.HRESULT(-((0x800B0100 ^ 0xFFFFFFFF) + 1))
+)
+
+// Verify performs authenticode verification on the file at path, and also
+// ensures that expectedCertSubject was the entity who signed it. path may point
+// to either a PE binary or an MSI package. ErrSigNotFound is returned if no
+// signature is found.
+func Verify(path string, expectedCertSubject string) error {
```

Comment:
This is a valid point. I definitely want to be able to extract cert subject for diagnostic reasons, however for verification it would be nice to use something better. OTOH, this code will be used by an older version desiring to upgrade to a new version, so it won't have awareness of new serials.

Acknowledging that my understanding of concrete implementations of PKI is somewhat limited, any suggestions/advice offered here would be useful.